### PR TITLE
Add parsing to some single value aggregations

### PR DESCRIPTION
**Pull request by cbuescher**
Similar to #23973 this adds parsing from xContent for the Min, Max, Avg, Sum and ValueCount aggregation.
I tried to pull out common functionality to ParsedNumericMetricsAggregation.SingleValue. Since InternalCardinality 
and InternalValueCount store a long value instead of a double, parsing and getters/setters for the value are slightly 
different so I didn't include their parsed counterparts under ParsedNumericMetricsAggregation.SingleValue for now.
Also included one case where I tried pulling out common xContent rendering into the shared interface (in Sum) for discussion.
I'm unsure if I like that one but it would help reduce some code duplication.

This PR is WIP against a feature branch.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/max/ParsedMax.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/max/ParsedMax.java]: Can we move all statics down? (after doXContentBody())

**Review by tlrx — COMMENTED** [review]
I had a first look, I'm happy to see that many things can be mutualized. I don't really like how the parsed Cardinality/ValueCount extends the ParsedNumericMetricsAggregation.SingleValue and don't use the `double` there and provide their own `long` value. I'm wondering if we could let SingleValue take care of the stringified value and have a Long and Double version of SingleValue where all parsing logic and XContent rendering would be placed.

I have to admit that I'm not sure if that's possible at all, so I'll take a look at this too

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: do we need this class? and why does SingleValue have to be an inner class?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: should we try and do without depending on DocValueFormat here? We can just do what this method does?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java]: I don't have a strong opinion here, I don't think it makes a whole lot of difference

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java]: where is this logic in InternalAvg?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java]: what's left todo here? the static method is already in the interface?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java]: here too, shall we try and not depend on the formatter?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java]: s/formated/formatted

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java]: leftover

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java]: complete POSITIVE with POSITIVE_INFINITY ?

**Review by javanna — APPROVED** [review]
left a few comments but LGTM

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java]: Will according to suggestion by @tlrx than, since I also don't mind much.

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: okay

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: okay

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: sure

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/max/ParsedMax.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/max/ParsedMax.java]: done

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java]: I changed this to Long.toString(valueCount).

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java]: yes

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java]: yes

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java]: done

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: I replaced this with Double.toString(value).

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: I don't know, the idea was to pull out value()/getValueAsString() and the internal fields for the Min/Max/Sum/Avg aggregation, maybe some others of the single value aggregations that just store a double. I tried to mimic the inheritance structure of InternalNumericMetricsAggregation here, which contains inner classes for SingleValue and MultiValue. Maybe we don't need it, but I'd like to keep it for now and revisit this decision once we are done with all the single value aggregations.

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java]: In InternalAvg the `value` field is only rendered if the avg normalizer (count) is not 0. (`count != 0 ? getValue() : null`). I choose to parse back `null` as Double.POSITIVE_INFINITY because that is what you get back when you divide a positive double by `0L`. So I check for that value here to get the same xContent output like InternalAvg.

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java]: This was an example of how we could share the xContent rendering logic between InternalXY and ParsedXY aggregations. I don't know if it would work in all places. I only did it for SUM, I could try to do it for Min/Max/Avg etc but that means slight changes in the existing internal aggregations. Not sure which way to go here atm, maybe you have some opinion?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: I think this hierarchy makes little sense and we should not copy it to our objects unless there are good reasons to do so. Is the only reason consistency?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java]: sounds good! leave a comment?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java]: I think we should do it only when it's very simple to do. It is a bit funky to have this static method in the interface but that is the best place for it. @tlrx what do you think?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: Is sharing value/valueAsString() a good reason to keep ParsedNumericMetricsAggregation? Or rename it to ParsedSingleValueMetricsAggregation and remove the inner SingleValue class? My suspicion at this point is that we might also reuse some stuff for the MultiValue case but I cannot tell until we start working on those.

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java]: I added a comment.

**Comment by cbuescher** [comment]
@tlrx @javanna thanks for the reviews, I added changes according to most of your comments, left two discussions open for now before I squash/merge this later

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java]: I do not see what ParsedNumericMetricsAggregation holds. Seems like it is used only as an ancestor for its inner class? In that case let's have ParsedSingleValueNumericMetricsAggregation as a top level class?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java]: I tried doing this for the aggregations in this PR in	1048678, but I'm not sure I like it. It avoids some code duplication but makes reading and reasonign about the code much harder in my opinion. Maybe we should only do it when it really saves a lot of duplication (like really long xContent methods), wdyt?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java]: yea I tend to agree, I am +0 on this :) I defer to Tanguy

**Comment by tlrx** [comment]
> Maybe we should only do it when it really saves a lot of duplication (like really long xContent methods), wdyt?

Yes, I agree. I think we should only do it when it saves a lot of duplication with custom logic. In this case it saves 1 line but added many more, and it also "force" people to look at what the Sum.renderXContent() is doing. I think we should not add those.

**Comment by cbuescher** [comment]
@tlrx thanks, I reverted the commit where I pulled the xContent rendering into the interfaces, rebased and reordered the PR so changes are more self contained. I think I will merge this one manually to the feature branch without squashing so that the history better reflects the individual changes to each aggregation.

**Pull request merged**
